book/src/rfcs/RFC0006-Dynamic-Databases.md

```diff
@@ -0,0 +1,635 @@
+  QueryGroupA -- depends on --> SalsaCrate;
+  QueryGroupB -- depends on --> SalsaCrate;
+```
+
+The result is that we do not actually compile a good part of the code from
+`QueryGroupA` or `QueryGroupB` until we build the final database crate.
+
+### What you can do today: dyn traits
+
+What you can do today is to use define a "dyn-compatible" query group
+trait and then write your derived functions using a `dyn` type as the
+argument:
+
+```rust,ignore
+#[salsa::query_group(QueryGroupAStorage)]
+trait QueryGroupA {
+    fn derived(&self, key: usize) -> usize;
+}
+
+fn derived(db: &dyn QueryGroupA, key: usize) -> usize {
+    key * 2
+}
+```
+
+This has the benefit that the `derived` function is not generic. However, it's
+still true that the glue code salsa makes will be generic over a `DB` type --
+this includes the impl of `QueryGroupA` but also the `Slot` and other machinery.
+This means that even if the only change is to query group B, in a different
+crate, the glue code for query group A ultimately has to be recompiled whenever
+the `Database` crate is rebuilt (though incremental compilation may help here).
+Moreover, as reported in [salsa-rs/salsa#220], measurements of rust-analyzer
+suggest that this code may be duplicated and accounting for more of the binary
+than we would expect.
+
+FIXME: I'd like to have better measurements on the above!
+
+[salsa-rs/salsa#220]: https://github.com/salsa-rs/salsa/issues/220
+
+### Our goal
+
+The primary goal of this RFC is to make it so that the glue code we generate for
+query groups is not dependent on the database type, thus enabling better
+incremental rebuilds.
+
+## User's guide
+
+Most of the changes in this RFC are "under the hood". But there are various user
+visibile changes proposed here. 
+
+### All query groups must be dyn safe
+
+The largest one is that **all Salsa query groups must now be dyn-safe**. The
+existing salsa query methods are all dyn-safe, so what this really implies is
+that one cannot have super-traits that use generic methods or other things that
+are not dyn safe. For example, this query group would be illegal:
+
+```rust,ignore
+#[salsa::query_group(QueryGroupAStorage)]
+trait QueryGroupA: Foo {
+}
+
+trait Foo {
+    fn method<T>(t: T) { }
+}
+```
+
+We could support query groups that are not dyn safe, but it would require us to
+have two "similar but different" ways of generating plumbing, and I'm not
+convinced that it's worth it. Moreover, it would require some form of opt-in so
+that would be a measure of user complexity as well.
+
+### The query method on `salsa::Database` moves to a helper method
+
+As a consequence of the previous point, the existing `query` and `query_mut`
+methods on the `salsa::Database` trait will be moved to an extension trait,
+automatically implemented for all salsa databases:
+
+```rust,ignore
+pub trait DatabaseExt {
+    #[allow(unused_variables)]
+    fn query<Q>(&self, query: Q) -> QueryTable<'_, Self, Q>
```

Comment:
Ah, interesting, that's true, we could do that, though it seems strictly less good. This is an interesting case in the language, it'd be nice if we had a way to express "final" functions in a trait that could not be overridden by impls (which would then be exempted from object safety requirements).

Reply:
In the latest version, I would up with `TheQuery.in_db(&db)` instead. This works better because we need to coerce the `&db` type to a `&dyn DB` anyway. It's also nicely ergonomic and avoids turbofish. Not as discoverable. We could, side note, add a method like `db.xxx_query()` that returns the `QueryTable`.